### PR TITLE
Preliminary support for Tenant scope resources

### PR DESCRIPTION
**Pull request by matthchr**
**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains tests

### CODE REVIEW

File: v2/pkg/genruntime/base_types.go [hunk 1]
```diff
@@ -16,9 +16,11 @@ type ResourceKind string
 
 const (
 	// ResourceKindNormal is a standard ARM resource.
-	ResourceKindNormal = ResourceKind("normal")
+	ResourceKindNormal = ResourceKind("normal") // TODO: Rename this to ResourceGroup?
```
Comment (theunrepentantgeek) [inline v2/pkg/genruntime/base_types.go:19]: Suggest doing this rename sooner rather than later, maybe even in this PR.

Comment (theunrepentantgeek) [reply v2/pkg/genruntime/base_types.go:19]: Though, is it the kind of resource that changes, or the kind of ownership?

File: v2/pkg/genruntime/root_resources.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ */
+
+package genruntime
+
+// TODO: The generated types should impl this
+type LocatableResource interface {
+	Location() string
+}
+
+// TenantResource is amrker interface that indicates that the implementing resource is scoped at the tenant level.
```
Comment (theunrepentantgeek) [inline v2/pkg/genruntime/root_resources.go:13]:
```suggestion
// TenantResource is a marker interface that indicates that the implementing resource is scoped at the tenant level.
```

File: v2/pkg/genruntime/root_resources.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package genruntime
+
+// TODO: The generated types should impl this
```
Comment (theunrepentantgeek) [inline v2/pkg/genruntime/root_resources.go:8]: Make an issue for this so we track it and come back to complete.

File: v2/tools/generator/internal/astmodel/resource_type.go [hunk 4]
```diff
@@ -18,13 +18,16 @@ import (
 	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astbuilder"
 )
 
+// ResourceKind is the scope the resource is deployed at.
 type ResourceKind string
```
Comment (theunrepentantgeek) [inline v2/tools/generator/internal/astmodel/resource_type.go:22]: Should this then be named `ResourceScope` instead?

File: v2/pkg/genruntime/root_resources.go (hunk 3 above)
Comment (theunrepentantgeek) [reply v2/pkg/genruntime/root_resources.go:8]: Or is this done already?

Comment (matthchr) [reply v2/pkg/genruntime/root_resources.go:8]: It's currently manually implemented on the one location-based resource we have (which is `ResourceGroup`)

Comment (matthchr) [reply v2/pkg/genruntime/root_resources.go:8]: but that does mean that there isn't really anything to do right now, as no resources actually require to to be automatically generated onto them

File: v2/pkg/genruntime/base_types.go (hunk 1 above)
Comment (matthchr) [reply v2/pkg/genruntime/base_types.go:19]: It's the scope of the resource that this really is talking about. I've updated this in this PR (although it does make this PR quite a bit bigger)

File: v2/tools/generator/internal/astmodel/resource_type.go (hunk 4 above)
Comment (matthchr) [reply v2/tools/generator/internal/astmodel/resource_type.go:22]: I've changed this name (and the corresponding one in genruntime) to scope, as that more closely reflects what it actually represents.

File: v2/pkg/genruntime/root_resources.go (hunk 3 above)
Comment (matthchr) [reply v2/pkg/genruntime/root_resources.go]: Filed #2420

**Review by theunrepentantgeek — APPROVED** [review]

**Pull request merged**